models/Asset.php

```diff
@@ -1540,68 +1540,88 @@ public function getMetadata($name = null, $language = null, $strictMatch = false
 
-        $convert = function ($metaData) {
-            $loader = Pimcore::getContainer()->get('pimcore.implementation_loader.asset.metadata.data');
-            $transformedData = $metaData['data'];
+        if ($name) {
+            return $this->getMetadataByName($name, $language, $strictMatch, $raw);
+        }
 
-            try {
-                /** @var Data $instance */
-                $instance = $loader->build($metaData['type']);
-                $transformedData = $instance->transformGetterData($metaData['data'], $metaData);
-            } catch (UnsupportedException $e) {
-            }
+        $metaData = $this->getObjectVar('metadata');
+        $result = [];
+        $metaDataWithLanguage = [];
+        if (is_array($metaData)) {
+            foreach ($metaData as $md) {
+                $md = (array)$md;
 
-            return $transformedData;
-        };
+                if ((empty($md['language']) && !$strictMatch) || ($language == $md['language']) || !$language) {
+                    if (!$raw) {
+                        $md['data'] = $this->transformMetadata($md);
+                    }
+                    $result[] = $md;
+                }
 
-        if ($name) {
-            if ($language === null) {
-                $language = Pimcore::getContainer()->get(LocaleServiceInterface::class)->findLocale();
+                if(!empty($md['language'])) {
+                    $metaDataWithLanguage[$md['language']][$md['name']] = $md;
+                }
             }
+        }
 
-            $data = null;
-            foreach ($this->metadata as $md) {
-                if ($md['name'] == $name) {
-                    if ($language == $md['language']) {
-                        if ($raw) {
-                            return $md;
-                        }
+        foreach($result as $key => &$item) {
+            if ($language && !$strictMatch) {
+                $lang = $item['language'];
 
-                        return $convert($md);
-                    }
-                    if (empty($md['language']) && !$strictMatch) {
-                        if ($raw) {
-                            return $md;
-                        }
-                        $data = $md;
+                if (!$lang && isset($metaDataWithLanguage[$language]) && isset($metaDataWithLanguage[$language][$item['name']])) {
+                    $itemWithLanguage = $metaDataWithLanguage[$language][$item['name']];
+                    if(!in_array($itemWithLanguage, $result)) {
+                        $item = $itemWithLanguage;
+                    } else {
+                        unset($result[$key]);
                     }
+
                 }
             }
+        }
 
-            if ($data) {
-                if ($raw) {
-                    return $data;
-                }
+        return $result;
+    }
 
-                return $convert($data);
-            }
+    protected function transformMetadata(array $metaData) {
```

Comment:
```suggestion
    private function transformMetadata(array $metaData) {
```
it was formerly an anonymous function `$convert`, probably it should be `private`, but i am not sure